db/migrate/20181026154056_add_type_to_users.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class AddTypeToUsers < ActiveRecord::Migration[5.2]
+  def up
+    # It's fine to set not null column this way, since we don't have tons of records yet
+    add_column :users, :type, :string, null: false, default: 'Customer'
+
+    execute <<-SQL
+      UPDATE users
+      SET type = CASE
+                   WHEN roles.name = 'root' THEN 'Root'
+                   WHEN roles.name = 'admin' THEN 'Admin'
+                   ELSE 'Customer'
+                 END
+      FROM roles
+      INNER JOIN users_roles ON users_roles.role_id = roles.id
+      WHERE users_roles.user_id = users.id AND users_roles.role_id = roles.id;
+    SQL
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration
```

Comment:
Irreversible migration? Seriously?

Reply:
Yeah, I did it on purpose. We should not let anyone to rollback this migration, since it's updated, based on the values from users_roles and roles tables. And we are removing them in next migration. OF COURSE, I can write SQL query in next migration (related to dropping users_roles and roles tables) and update roles.name according to `type` attribute in the users table, but I won't do it, because:
1. It was bad decision to have users_roles table, so we won't support it anymore;
2. Anyone still will be able to rollback this migration by simply `rake db:migrate:down VERSION=`, so we will lost data about user's types. We can also handle it by creating backup, but I don't think, that it's necessary for this pet project :bowtie:
3. Making it dependent from the any other migration bad practice.
4. It's okay to do that if you are sure that your migration should not be rolled back. Check this https://edgeguides.rubyonrails.org/active_record_migrations.html#using-reversible
`Sometimes your migration will do something which is just plain irreversible; for example, it might destroy some data. In such cases, you can raise ActiveRecord::IrreversibleMigration in your down block. If someone tries to revert your migration, an error message will be displayed saying that it can't be done.`

Reply:
I think that these two migrations should be merged in one.
If you were able to save the data during this migration, why do you think it's unreal to do it in reverse, at least as a challenge for yourself?

If you have no time for this, let me know, please, I will do it later then.

Reply:
Agree with merging! No, it's fine, I have a time, no worry :smiley: